File: frontend/src/services/bookmarkService.ts
import api from './api'

export interface BookmarkDto {
  id: number
  userId: string
  drawNumber: number
  label: string
  description: string
  createdAt: string
  updatedAt?: string
  drawDate: string
  winningNumbers: number[]
  draw?: {
    draw: number
    date: string
    winningNumber1: number
    winningNumber2: number
    winningNumber3: number
    winningNumber4: number
    winningNumber5: number
    winningNumber6: number
    bonusNumber?: number
    powerballNumber?: number
  }
}

export interface CreateBookmarkRequest {
  drawNumber: number
  label: string
  description?: string
}

export interface UpdateBookmarkRequest {
  label: string
  description?: string
}

export class BookmarkService {
  private readonly baseUrl = '/bookmarks'

  async createBookmark(request: CreateBookmarkRequest): Promise<BookmarkDto> {
    const response = await api.post<BookmarkDto>(this.baseUrl, request)
    return response.data
  }

  async getUserBookmarks(): Promise<BookmarkDto[]> {
    const response = await api.get<BookmarkDto[]>(this.baseUrl)
    return response.data
  }

  async updateBookmark(bookmarkId: number, request: UpdateBookmarkRequest): Promise<BookmarkDto> {
    const response = await api.put<BookmarkDto>(`${this.baseUrl}/${bookmarkId}`, request)
    return response.data
  }

  async deleteBookmark(bookmarkId: number): Promise<boolean> {
    try {
      await api.delete(`${this.baseUrl}/${bookmarkId}`)
      return true
    } catch (error) {
      console.error('Error deleting bookmark:', error)
      return false
    }
  }

  async bookmarkExists(drawNumber: number): Promise<boolean> {
    try {
      const response = await api.get<boolean>(`${this.baseUrl}/exists/${drawNumber}`)
      return response.data
    } catch (error) {
      console.error('Error checking bookmark existence:', error)
      return false
    }
  }

  async reorderBookmarks(bookmarkIds: number[]): Promise<boolean> {
    try {
      await api.post(`${this.baseUrl}/reorder`, { bookmarkIds })
      return true
    } catch (error) {
      console.error('Error reordering bookmarks:', error)
      return false
    }
  }
}

export const bookmarkService = new BookmarkService()

File: backend/Services/BookmarkService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using PredictLottoNZ.Data;
using PredictLottoNZ.Models;
using PredictLottoNZ.Models.DTOs;

namespace PredictLottoNZ.Services;

public class BookmarkService : IBookmarkService
{
    private readonly LottoDbContext _context;
    private readonly ILogger<BookmarkService> _logger;

    public BookmarkService(LottoDbContext context, ILogger<BookmarkService> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<BookmarkDto> CreateBookmarkAsync(int drawNumber, string label, string userId)
    {
        if (drawNumber <= 0)
        {
            throw new ArgumentException("Draw number must be positive", nameof(drawNumber));
        }

        if (string.IsNullOrWhiteSpace(label))
        {
            throw new ArgumentException("Label cannot be null or empty", nameof(label));
        }

        if (string.IsNullOrWhiteSpace(userId))
        {
            throw new ArgumentException("User ID cannot be null or empty", nameof(userId));
        }

        // Check if draw exists
        var drawExists = await _context.LottoDraws.AnyAsync(d => d.Draw == drawNumber);
        if (!drawExists)
        {
            throw new ArgumentException($"Draw {drawNumber} does not exist", nameof(drawNumber));
        }

        // Check if bookmark already exists for this user and draw
        var existingBookmark = await _context.Bookmarks
            .FirstOrDefaultAsync(b => b.UserId == userId && b.DrawNumber == drawNumber);

        if (existingBookmark != null)
        {
            throw new InvalidOperationException($"Bookmark already exists for draw {drawNumber} and user {userId}");
        }

        var bookmark = new Bookmark
        {
            UserId = userId,
            DrawNumber = drawNumber,
            Label = label.Trim(),
            Description = string.Empty,
            CreatedAt = DateTime.UtcNow
        };

        _context.Bookmarks.Add(bookmark);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Created bookmark {BookmarkId} for draw {DrawNumber} by user {UserId}", 
            bookmark.Id, drawNumber, userId);

        return new BookmarkDto
        {
            Id = bookmark.Id,
            DrawNumber = bookmark.DrawNumber,
            Label = bookmark.Label,
            Description = bookmark.Description,
            CreatedAt = bookmark.CreatedAt,
            UpdatedAt = bookmark.UpdatedAt
        };
    }

    public async Task<IEnumerable<BookmarkDto>> GetUserBookmarksAsync(string userId)
    {
        if (string.IsNullOrWhiteSpace(userId))
        {
            throw new ArgumentException("User ID cannot be null or empty", nameof(userId));
        }

        var bookmarks = await _context.Bookmarks
            .Where(b => b.UserId == userId)
            .Include(b => b.Draw)
            .OrderByDescending(b => b.CreatedAt)
            .Select(b => new BookmarkDto
            {
                Id = b.Id,
                DrawNumber = b.DrawNumber,
                Label = b.Label,
                Description = b.Description,
                CreatedAt = b.CreatedAt,
                UpdatedAt = b.UpdatedAt,
                DrawDate = b.Draw != null ? b.Draw.Date : DateTime.MinValue,
                WinningNumbers = b.Draw != null ? new int[] 
                { 
                    b.Draw.WinningNumber1, 
                    b.Draw.WinningNumber2, 
                    b.Draw.WinningNumber3, 
                    b.Draw.WinningNumber4, 
                    b.Draw.WinningNumber5, 
                    b.Draw.WinningNumber6 
                } : Array.Empty<int>()
            })
            .ToListAsync();

        _logger.LogInformation("Retrieved {Count} bookmarks for user {UserId}", bookmarks.Count, userId);
        return bookmarks;
    }

    public async Task<bool> DeleteBookmarkAsync(int bookmarkId, string userId)
    {
        if (bookmarkId <= 0)
        {
            throw new ArgumentException("Bookmark ID must be positive", nameof(bookmarkId));
        }

        if (string.IsNullOrWhiteSpace(userId))
        {
            throw new ArgumentException("User ID cannot be null or empty", nameof(userId));
        }

        var bookmark = await _context.Bookmarks
            .FirstOrDefaultAsync(b => b.Id == bookmarkId && b.UserId == userId);

        if (bookmark == null)
        {
            _logger.LogWarning("Bookmark {BookmarkId} not found for user {UserId}", bookmarkId, userId);
            return false;
        }

        _context.Bookmarks.Remove(bookmark);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Deleted bookmark {BookmarkId} for draw {DrawNumber} by user {UserId}", 
            bookmarkId, bookmark.DrawNumber, userId);

        return true;
    }

    public async Task<bool> BookmarkExistsAsync(string userId, int drawNumber)
    {
        if (string.IsNullOrWhiteSpace(userId))
        {
            throw new ArgumentException("User ID cannot be null or empty", nameof(userId));
        }

        if (drawNumber <= 0)
        {
            throw new ArgumentException("Draw number must be positive", nameof(drawNumber));
        }

        return await _context.Bookmarks
            .AnyAsync(b => b.UserId == userId && b.DrawNumber == drawNumber);
    }

    public async Task<BookmarkDto?> UpdateBookmarkAsync(int bookmarkId, string newLabel, string userId)
    {
        if (bookmarkId <= 0)
        {
            throw new ArgumentException("Bookmark ID must be positive", nameof(bookmarkId));
        }

        if (string.IsNullOrWhiteSpace(newLabel))
        {
            throw new ArgumentException("New label cannot be null or empty", nameof(newLabel));
        }

        if (string.IsNullOrWhiteSpace(userId))
        {
            throw new ArgumentException("User ID cannot be null or empty", nameof(userId));
        }

        var bookmark = await _context.Bookmarks
            .FirstOrDefaultAsync(b => b.Id == bookmarkId && b.UserId == userId);

        if (bookmark == null)
        {
            _logger.LogWarning("Bookmark {BookmarkId} not found for user {UserId}", bookmarkId, userId);
            return null;
        }

        bookmark.Label = newLabel.Trim();
        bookmark.UpdatedAt = DateTime.UtcNow;

        await _context.SaveChangesAsync();

        _logger.LogInformation("Updated bookmark {BookmarkId} label to '{NewLabel}' for user {UserId}", 
            bookmarkId, newLabel, userId);

        return new BookmarkDto
        {
            Id = bookmark.Id,
            DrawNumber = bookmark.DrawNumber,
            Label = bookmark.Label,
            Description = bookmark.Description,
            CreatedAt = bookmark.CreatedAt,
            UpdatedAt = bookmark.UpdatedAt
        };
    }
}
